native/cocos/platform/SDLHelper.cpp

```diff
@@ -0,0 +1,377 @@
+            _delegate->dispatchEvent(touch);
+            break;
+        }
+        case SDL_FINGERMOTION: {
+            const SDL_TouchFingerEvent &event = sdlEvent.tfinger;
+            touch.type                        = TouchEvent::Type::MOVED;
+            touch.touches                     = {TouchInfo(event.x, event.y, (int)event.fingerId)};
+            _delegate->dispatchEvent(touch);
+            break;
+        }
+        case SDL_KEYDOWN: {
+            const SDL_KeyboardEvent &event = sdlEvent.key;
+            auto                     mode  = event.keysym.mod;
+            keyboard.action                = KeyboardEvent::Action::PRESS;
+            keyboard.key                   = sdlKeycodeToCocosCode(event.keysym.sym, mode);
+            keyboard.altKeyActive          = mode & KMOD_ALT;
+            keyboard.ctrlKeyActive         = mode & KMOD_CTRL;
+            keyboard.shiftKeyActive        = mode & KMOD_SHIFT;
+            //CC_LOG_DEBUG("==> key %d -> code %d", event.keysym.sym, keyboard.key);
+            _delegate->dispatchEvent(keyboard);
+            break;
+        }
+        case SDL_KEYUP: {
+            const SDL_KeyboardEvent &event = sdlEvent.key;
+            auto                     mode  = event.keysym.mod;
+            keyboard.action                = KeyboardEvent::Action::RELEASE;
+            keyboard.key                   = sdlKeycodeToCocosCode(event.keysym.sym, mode);
+            keyboard.altKeyActive          = mode & KMOD_ALT;
+            keyboard.ctrlKeyActive         = mode & KMOD_CTRL;
+            keyboard.shiftKeyActive        = mode & KMOD_SHIFT;
+            _delegate->dispatchEvent(keyboard);
+            break;
+        }
+        default:
+            break;
+    }
+}
+
+void SDLHelper::pollEvent(bool *quit) {
+    SDL_Event sdlEvent;
+    int       cnt = 0;
+    while ((cnt = SDL_PollEvent(&sdlEvent)) != 0) {
+        dispatchSDLEvent(sdlEvent, quit);
+    }
+}
+
+bool SDLHelper::createWindow(const char *title,
+                             int x, int y, int w,
+                             int h, int flags) {
+    // Create window
+    int sdlFlags = windowFlagsToSDLWindowFlag(flags);
+    _handle      = SDL_CreateWindow(title, SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED, w, h, sdlFlags);
+    if (_handle == nullptr) {
+        // Display error message
+        CC_LOG_ERROR("Window could not be created! SDL_Error: %s\n", SDL_GetError());
+        return false;
+    }
+    _windowCreated = true;
+    return true;
+}
+
+void SDLHelper::swapWindow() {
+    SDL_GL_SwapWindow(_handle);
+}
+
+#if (CC_PLATFORM == CC_PLATFORM_LINUX)
+uintptr_t SDLHelper::getDisplay() const {
+    SDL_SysWMinfo wmInfo;
+    SDL_VERSION(&wmInfo.version);
+    SDL_GetWindowWMInfo(_handle, &wmInfo);
+    return reinterpret_cast<uintptr_t>(wmInfo.info.x11.display);
+}
+#endif
+
+uintptr_t SDLHelper::getWindowHandler() const {
+    SDL_SysWMinfo wmInfo;
+    SDL_VERSION(&wmInfo.version);
```

Comment:
Is it necessary to invoke SDL_VERSION to get the version information here?

Reply:
It is needed. The internal implementation does different operations according to different versions
![image](https://user-images.githubusercontent.com/5205637/158938244-a470e35a-e5e2-4240-9ca3-1d42cff39fbc.png)